sources/fedora-http.go

```diff
@@ -173,3 +216,124 @@ func (s *fedora) getLatestBuild(URL, release string) (string, error) {
 	return matches[len(matches)-1], nil
 }
+
+// getDownloadURL fetches JSON representation of current releases and
+// extracts the download URL matching the given release, variant, arch,
+// and filetype (.tar.xz, .iso, etc.).
+func (s *fedora) getDownloadURL(release string, variant string, arch string, fileType string) (url string, checksum string, err error) {
+	releasesURL := "https://fedoraproject.org/releases.json"
+	s.logger.Infof("Querying %q for list of available releases", releasesURL)
+
+	// Fetch available releases.
+	resp, err := http.Get(releasesURL)
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to fetch releases from %q: %w", releasesURL, err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", "", fmt.Errorf("Unexpected status code %q from %q: %w", resp.StatusCode, releasesURL, err)
+	}
+
+	// Parse JSON response.
+	var result []struct {
+		Release    string `json:"version"`
+		Arch       string `json:"arch"`
+		Variant    string `json:"variant"`
+		Subvariant string `json:"subvariant"`
+		URL        string `json:"link"`
+		SHA256     string `json:"sha256"`
+		SizeBytes  string `json:"size"`
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&result)
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to parse JSON response: %w", err)
+	}
+
+	// Variant mapping.
+	// Default to variant "Container" and subvariant "Container_Base".
+	var subvariant string
+	if variant == "default" || variant == "cloud" {
+		variant = "Server"
+		subvariant = "Server"
+	}
+
+	// Iterate over response items and find the matching image.
+	for _, item := range result {
+		if item.Release != release || item.Arch != arch || item.Variant != variant {
+			continue
+		}
+
+		if subvariant != "" && item.Subvariant != subvariant {
+			continue
+		}
+
+		if !strings.HasSuffix(item.URL, fileType) {
+			continue
+		}
+
+		// Matching URL found.
+		return item.URL, "", nil
+	}
+
+	return "", "", fmt.Errorf("Failed to find download URL for release %q, variant %q, arch %q, and file type %q", release, variant, arch, fileType)
+}
+
+func (s *fedora) isoRunner(gpgkeys string) error {
+	err := shared.RunScript(s.ctx, fmt.Sprintf(`#!/bin/sh
+set -eux
+
+# Create required files
+touch /etc/mtab /etc/fstab
+
+dnf_args=""
+mkdir -p /etc/yum.repos.d
+
+# Add cdrom repo
+cat <<- EOF > /etc/yum.repos.d/cdrom.repo
+[cdrom]
+name=Install CD-ROM
+baseurl=file:///mnt/cdrom
+enabled=0
+EOF
+
+GPG_KEYS="%s"
+
+gpg_keys_official="file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-%[2]s-primary"
+
+if [ -n "${GPG_KEYS}" ]; then
+	echo gpgcheck=1 >> /etc/yum.repos.d/cdrom.repo
+	echo gpgkey=${gpg_keys_official} ${GPG_KEYS} >> /etc/yum.repos.d/cdrom.repo
+else
+	echo gpgcheck=0 >> /etc/yum.repos.d/cdrom.repo
+fi
+
+dnf_args="--disablerepo=* --enablerepo=cdrom"
+
+# Newest install.img doesnt have rpm installed,
+# so first install rpm.
+if [ -z "$(which rpmkeys)" ]; then
+	cd /mnt/cdrom/Packages
+	dnf ${dnf_args} -y install rpm --nogpgcheck
```

Comment:
Should that `--nogpgcheck` be conditional to having no actual GPG keys to check with?